extra/tensors/tensors.factor

```diff
@@ -0,0 +1,257 @@
+    ! For each element in the range, we want to compute the dot product of the
+    ! corresponding row and column
+    res
+    [   >fixnum
+        ! Get the row
+        [ [ vec1 n ] dip p row ]
+        ! Get the column
+        ! [ p mod vec2 swap p every ] bi
+        [ p mod f p vec2 <step-slice> ] bi
+        ! Take the dot product
+        [ * ] [ + ] 2map-reduce
+    ]
+    map! drop ;
+
+PRIVATE>
+
+
+! Perform matrix multiplication muliplying an
+! ...xmxn matrix with a ...xnxp matrix
+TYPED:: matmul ( tensor1: tensor tensor2: tensor -- tensor3: tensor )
+    ! First check the shape
+    tensor1 tensor2 check-matmul-shape
+
+    ! Now save all of the sizes
+    tensor1 shape>> unclip-last-slice :> n
+    unclip-last-slice :> m :> top-shape
+    tensor2 shape>> last :> p
+    top-shape product :> rest
+
+    ! Now create the new tensor with { 0 ... m*p-1 } repeating
+    top-shape { m p } append naturals m p * t% :> tensor3
+
+    ! Now update the tensor3 to contain the multiplied matricies
+    rest [0,b)
+    [
+        :> i
+        ! First make vec1
+        m n * i * dup m n * + tensor1 vec>> <slice>
+        ! Now make vec2
+        n p * i * dup n p * + tensor2 vec>> <slice>
+        ! Now make the resulting vector
+        m p * i * dup m p * + tensor3 vec>> <slice>
+        ! Push n and p and multiply the clices
+        n p 2d-matmul
+        0
+    ] map drop
+    tensor3 ;
+
+<PRIVATE
+! helper for transpose: gets the turns a shape into a list of things
+! by which to multiply indices to get a full index
+: get-mults ( shape -- seq )
+    ! dup [ 1 0 ] dip set-nth ! set 1st num to 1
+    rest reverse { 1 } swap append
+    1 swap ! put accumulator under shape on stack
+    [ * dup ] map ! finds all the nice values
+    [ drop ] dip ; ! remove accumulator
+
+! helper for transpose: given shape & flat index, gives nd index for transpose
+:: trans-index ( ind shape -- seq )
+    ! what we use to divide things
+    shape reverse :> S
+    ! what we multiply by
+    shape get-mults :> mults
+    ! accumulator
+    { }
+    ! loop thru amnts we want to slice off of S
+    0 [ dup S length < ]
+    [
+        ! we divide by the product of the 1st n elements of S
+        dup S swap head-slice product :> div
+        ! we mod by the nth element of S
+        dup S nth :> m
+        ! BUT: we do not mod on the last index
+        dup S length 1 - = not :> mod?
+        [ ! dip to ignore iterator
+            ! multiply accumulator by mults & sum
+            dup mults [ * ] 2map sum
+            ! subtract from ind & divide
+            ind swap - div /
+            ! mod if necessary
+            mod? [ m mod ] [ ] if
+            ! append to accumulator
+            1array append
+        ] dip
+        ! increment iterator
+        1 +
+    ] while drop ! drop iterator
+    reverse ;
+PRIVATE>
+
+! Transpose an n-dimensional tensor
+TYPED:: transpose ( tensor: tensor -- tensor': tensor )
```

Comment:
Have you done any benchmarking or profiling of the code to see how it performs roughly against something like ``flip`` for a two-dimensional array, or against ``ndarray.T``?

Reply:
No benchmarking has been done yet. That would be a goal for next semester.